File: 백준/Gold/15486. 퇴사 2/퇴사 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int N;
    static int[][] arr;
    static int[] dp;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();


        N = Integer.parseInt(br.readLine());
        arr = new int[N][2];
        dp = new int[N + 1];
        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());

            arr[i][0] = Integer.parseInt(st.nextToken());
            arr[i][1] = Integer.parseInt(st.nextToken());
        }

        dp[N] = 0;
        for (int i = N - 1; i >= 0; i--) {
            if (i + arr[i][0] < N + 1) {
                dp[i] = Math.max(dp[i + 1], dp[i + arr[i][0]] + arr[i][1]);
            } else {
                dp[i] = dp[i + 1];
            }
        }
        System.out.println(dp[0]);
    }

    public static int recur(int time) {
        if (time > N) {
            return -99999999;
        }

        if (time == N) {
            return 0;
        }

        if (dp[time] != -1) {
            return dp[time];
        }


        return dp[time] = Math.max(recur(time + 1), recur(time + arr[time][0]) + arr[time][1]);
    }


}

File: 백준/Gold/15486. 퇴사 2/퇴사 2.py
import sys
# sys.setrecursionlimit(10**6)
n = int(sys.stdin.readline())
arr = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]

# 탑다운
dp = [-1234567890] * 1500051
# dp[i]: i번째 날부터 상담을 시작했을 때 얻을 수 있는 최대 이익
dp[n] = 0
for i in range(n - 1, -1, -1):
    dp[i] = max(dp[i + arr[i][0]] + arr[i][1], dp[i + 1])
print(dp[0])